Automattic/jetpack
# Reconnect Process: Partial Reconnect

~Requires `D46648-code` and `D47087-code`  patches to be applied.~ Patches deployed.

#### Changes proposed in this Pull Request:
- Blog and user tokens get validated via the `jetpack-token-health` WP.com endpoint.
- If the user token is invalid, it gets refreshed.
- If the blog token is invalid, it gets refreshed via the  `jetpack-refresh-blog-token` WP.com endpoint.

#### Jetpack product discussion
`p9dueE-1CV-p2`

#### Does this pull request change what data or activity we track or use?
Yes.
It adds three new events for the "Restore Connection" process:
- `restore_connection_reconnect` (full reconnect started)
- `restore_connection_refresh_blog_token` (blog token refresh initiated)
- `restore_connection_refresh_user_token` (user token refresh initiated)
Any of these events can only be triggered after Jetpack has already been connected, therefore TOS have been accepted.
Additionally, there's [TOS check](https://github.com/Automattic/jetpack/blob/master/packages/tracking/src/class-tracking.php#L107) in the code that doesn't allow the events to be tracked if TOS wasn't accepted.

#### Testing instructions:
1. Check out the branch, install the "Debug Tools", activate the "Broken Token".
2. Break the blog token.
3. Go to Jetpack Dashboard and confirm that you see the error notice.
4. Click "Restore Connection" button.
5. The dashboard should reload, the blog token should be restored. You *should not* have to click on the "Approve" button.
7. Break the user token, save somewhere existing blog token.
8. Add following code into your theme's `functions.php`:
```php
add_filter( 'react_connection_errors_initial_state', function( $errors ) {
	$errors[] = array(
		'message' => __('Connection Error'),
		'action' => 'reconnect',
		'code' => 'invalid_blog_token',
	);

	return $errors;
} );
```
9. Reload the Jetpack Dashboard, click "Restore Connection" in the error notice.
10. The authorization iframe should appear below the error notice, click "Approve".
11. Jetpack Dashboard should reload, the broken user token should be restored. The blog token _should not_ change.
12. **Do not** break any tokens, but open the "Broken Token" tab and save the current tokens somewhere.
13. Go to the Jetpack Dashboard, click "Restore Connection". Go through the authorization process.
14. Open another "Broken Token" tab and confirm that both blog and user tokens have changed.

#### Proposed changelog entry for your changes:
* Improved restoring connection to WordPress.com in case of connection errors.
